pump/server_test.go

```diff
@@ -555,3 +565,222 @@ func (s *listenSuite) TestReturnListener(c *C) {
+	c.Assert(p, IsNil)
+	c.Assert(err, ErrorMatches, "URL.*")
+}
+
+func (s *newServerSuite) TestCreateNewPumpServer(c *C) {
+	util.NewPdCli = func([]string, pd.SecurityOption) (pd.Client, error) {
+		return &mockPdCli{}, nil
+	}
+	newTiKVLockResolver = func([]string, config.Security) (*tikv.LockResolver, error) {
+		return nil, nil
+	}
+	newKVStore = func(path string) (kv.Storage, error) {
+		return nil, nil
+	}
+
+	p, err := NewServer(s.cfg)
+	c.Assert(err, IsNil)
+	c.Assert(p.clusterID, Equals, uint64(8012))
+	p.Close()
+}
+
+type startNode struct {
+	status *node.Status
+}
+
+func (n *startNode) ID() string                                                   { return "startnode-long" }
+func (n *startNode) ShortID() string                                              { return "startnode" }
+func (n *startNode) RefreshStatus(ctx context.Context, status *node.Status) error { return nil }
+func (n *startNode) Heartbeat(ctx context.Context) <-chan error                   { return make(chan error) }
+func (n *startNode) Notify(ctx context.Context) error                             { return nil }
+func (n *startNode) NodeStatus() *node.Status {
+	return n.status
+}
+func (n *startNode) NodesStatus(ctx context.Context) ([]*node.Status, error) {
+	return []*node.Status{n.status}, nil
+}
+func (n *startNode) Quit() error { return nil }
+
+type startServerSuite struct {
+	origUtilGetTSO func(pd.Client) (int64, error)
+}
+
+var _ = Suite(&startServerSuite{})
+
+func (s *startServerSuite) SetUpTest(c *C) {
+	s.origUtilGetTSO = utilGetTSO
+}
+
+func (s *startServerSuite) TearDownTest(c *C) {
+	utilGetTSO = s.origUtilGetTSO
+}
+
+func (s *startServerSuite) TestStartPumpServer(c *C) {
+	utilGetTSO = func(pd.Client) (int64, error) {
+		return 42, nil
+	}
+	etcdClient := testEtcdCluster.RandClient()
+	cfg := &Config{
+		ListenAddr:        "http://127.0.0.1:8250",
+		AdvertiseAddr:     "http://127.0.0.1:8260",
+		Socket:            "unix://127.0.0.1:8280/hello/world",
+		EtcdURLs:          strings.Join(etcdClient.Endpoints(), ","),
+		DataDir:           "/tmp/pump",
+		HeartbeatInterval: 1500,
+		LogLevel:          "debug",
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	grpcOpts := []grpc.ServerOption{grpc.MaxRecvMsgSize(GlobalConfig.maxMsgSize)}
+	p := &Server{
+		dataDir:    cfg.DataDir,
+		storage:    &noOpStorage{},
+		clusterID:  8012,
+		node:       &startNode{status: &node.Status{State: node.Online, NodeID: "startnode-long", Addr: "http://192.168.199.100:8260"}},
+		tcpAddr:    cfg.ListenAddr,
+		unixAddr:   cfg.Socket,
+		gs:         grpc.NewServer(grpcOpts...),
+		ctx:        ctx,
+		cancel:     cancel,
+		tiStore:    nil,
+		gcDuration: time.Duration(cfg.GC) * 24 * time.Hour,
+		pdCli:      nil,
+		cfg:        cfg,
+		triggerGC:  make(chan time.Time)}
+	defer p.Close()
+	go p.Start()
+
+	// wait until the server is online
+	timeEnd := time.After(5 * time.Second)
+	getInterval := time.Tick(500 * time.Microsecond)
```

Comment:
Use `NewTicker` instead so that we can stop it explicitly.

Reply:
Revised